eth/executionclient/execution_client.go

```diff
@@ -345,63 +457,73 @@ func (ec *ExecutionClient) streamLogsToChan(ctx context.Context, logs chan<- Blo
+			observability.RecordUint64Value(ctx, fromBlock, lastProcessedBlockGauge.Record, metric.WithAttributes(semconv.ServerAddress(mc.addr)))
 		}
 	}
 }
 
-// connect connects to Ethereum execution client.
-// It must not be called twice in parallel.
-func (ec *ExecutionClient) connect(ctx context.Context) error {
-	logger := ec.logger.With(fields.Address(ec.nodeAddr))
+// getHealthyClient selects the first healthy ManagedClient and returns it along with its ethclient.Client.
+func (ec *ExecutionClient) getHealthyClient() (*ManagedClient, *ethclient.Client, error) {
+	ec.clientsMu.RLock()
+	defer ec.clientsMu.RUnlock()
 
-	ctx, cancel := context.WithTimeout(ctx, ec.connectionTimeout)
-	defer cancel()
+	for _, mc := range ec.clients {
+		if mc.isHealthy() {
+			client := mc.getClient()
+			if client != nil {
+				return mc, client, nil
+			}
+		}
+	}
+	return nil, nil, ErrNotConnected
+}
 
-	start := time.Now()
-	var err error
-	ec.client, err = ethclient.DialContext(ctx, ec.nodeAddr)
+// Filterer returns a contract filterer using the first healthy client.
+func (ec *ExecutionClient) Filterer() (*contract.ContractFilterer, error) {
+	_, client, err := ec.getHealthyClient()
 	if err != nil {
-		ec.logger.Error(elResponseErrMsg,
-			zap.String("operation", "DialContext"),
-			zap.Error(err))
-		return err
+		return nil, err
 	}
-
-	logger.Info("connected to execution client", zap.Duration("took", time.Since(start)))
-	return nil
+	return contract.NewContractFilterer(ec.contractAddress, client)
 }
 
-// reconnect tries to reconnect multiple times with an exponent interval.
-// It panics when reconnecting limit is reached.
-// It must not be called twice in parallel.
-func (ec *ExecutionClient) reconnect(ctx context.Context) {
-	logger := ec.logger.With(fields.Address(ec.nodeAddr))
-
-	start := time.Now()
-	tasks.ExecWithInterval(func(lastTick time.Duration) (stop bool, cont bool) {
-		logger.Info("reconnecting")
-		if err := ec.connect(ctx); err != nil {
-			if ec.isClosed() {
-				return true, false
-			}
-			// continue until reaching to limit, and then panic as Ethereum execution client connection is required
-			if lastTick >= ec.reconnectionMaxInterval {
-				logger.Panic("failed to reconnect", zap.Error(err))
-			} else {
-				logger.Warn("could not reconnect, still trying", zap.Error(err))
-			}
-			return false, false
-		}
-		return true, false
-	}, ec.reconnectionInitialInterval, ec.reconnectionMaxInterval+(ec.reconnectionInitialInterval))
+type callFunc func(ctx context.Context, mc *ManagedClient) (any, error)
 
-	logger.Info("reconnected to execution client", zap.Duration("took", time.Since(start)))
-}
+// doCall carries out a call on the healthy clients until one succeeds.
+func (ec *ExecutionClient) doCall(ctx context.Context, call callFunc) (any, error) {
+	ec.clientsMu.RLock()
+	clients := make([]*ManagedClient, len(ec.clients))
+	copy(clients, ec.clients)
+	ec.clientsMu.RUnlock()
 
```

Comment:
is the `ec.client` changing? why do you need to protect it? how does this copy help if its all pointers?

Reply:
also why not just use `getHealthyClient` here?

Reply:
I just changed the implementation to use multi-client, so I'm resolving this